### ISSUE TITLE
 Recommendation/best practices for selecting learning_rate_warmup_steps wrt batch size.

### ISSUE THREAD

**Issue by billy19murahmi** [tensorflow/tensor2tensor#991]
Hey,

For Machine Translation (Transformer model) Default of 16k steps is what I'm using for learning_rate_warmup_steps irrespective of the batch size. 
Are there any recommendation or best practices for selecting learning_rate_warmup_steps based as we scale up the batch size?

Thanks,
Srini

**Comment by martinpopel** [comment]
Dear Srini,
see https://ufal.mff.cuni.cz/pbml/110/art-popel-bojar.pdf#page=21
It discusses the relation between effective batch size and LR & warmup steps.
It focuses on increasing the effective batch size with multiple GPU,
but there is relevant info also for singe-GPU settings.

Figure 8 shows that the effect of warmup is either negligible or crucial (in cases where it prevents divergence).
Figure 8 does not vary batch_size, but I have similar experience there.
My conclusions:
- If you have time&resources, you can tune warmups on a dev set, but should not expect any breakthrough.
- If you don't have time, just keep the default warmup but watch for divergence - if it happens increase the warmup
(and/or add gradient clipping and/or decrease LR).

Best,
Martin

----- Original Message -----

**Comment by billy19murahmi** [comment]
Hey Martin,

Thank you. I've read the paper. The study you guys have done is amazing and very helpful to understand the effect of those hyper-parameters.  Fig 8 shows the effect of warmup steps with respect to batch_size=1500. From the study it's clear about the influence of warmup steps, and we need to check for divergence and make changes accordingly. But i was wondering if we increase the effective batch size to 65k, should we scale down the warmup steps? because the model would see much data in fewer number of steps. So I was wondering if you've done such a study.

Thanks,
Srini

**Comment by martinpopel** [comment]
> increase the effective batch size to 65k

So you also use multiple GPUs? Or MultistepAdamOptimizer?

I have tried effective batch size just up to 8*2900 (with Adafactor), not more. 
In general: if you plan to train for e.g. one week, it makes no harm to use a warmup corresponding to a few hours.

**Comment by billy19murahmi** [comment]
I'm not using GPU's. I'm experimenting with CPU's.
I want to run it on multi-CPU cluster (Distributed training). Problem is, it takes much time to figure out and change the parameters due to large computation time.  So if any previous study is done with respect to warmup steps on higher effective batch_size, I can keep that as baseline and experiment further.

**Comment by billy19murahmi** [comment]
@vsuthichai yes I'm getting comparable translation results with effective large batches on CPU's. Used the recommendations from Facebook's "scaling neural machine translation" paper for the baseline experiments on large batches.

**Comment by vsuthichai** [comment]
@billy19murahmi What would be your time to train to convergence?

**Comment by billy19murahmi** [comment]
@vsuthichai Since tests are on cpu's it used to take few weeks on single node.
Distributing it on 50 nodes, it was able to converge in less than 3 days. 
It reaches 27.5-28 BLEU in few hours using more number of nodes. 
We're still trying to optimize it.
